sql/schemachanger: enable add column by default in declarative schema changer

fqazi opened the pull request:
These changes will enable add column by default in the declarative schema changer and the following fixes:

1. Add support for retrying schema changes when retriable errors are encountered
2. Add support for refreshing stats during declarative schema changes
3. Block add column when cross-database references are observed in the declarative schema changer
4. Block implicit record types inside expressions for the declarative schema changer
5. Add schema change-related event log entries
6. Properly combine multiple declarative schema change operations inside CTEs, which is required for support geometry related builtins that add columns
7. Enforce schema usage related privileges in the declarative schema changer
8. Add add column related telemetry
9. Generate appropriate errors when a newly added column conflicts with a system column
10. Support empty column family names
11. Add columns in the user-specified order inside the declarative schema changer
12. Enforce constraints when adding new columns, specifically detecting if default values will violate them

Along with the following known issues and limitations (tracked by issue: #80545)
1. Legacy schema changer will be used whenever serial and generated columns are observed.
2. Regional by row tables will default to the legacy schema changer since we are missing proper support for updating zone configs
3. Columns with the unique constraint will default to the legacy schema changer
4. Default expressions using sequences will default to the legacy schema changer
5. A number of tests cases depending on the MVCC backfiller are using the legacy schema changer for now

fqazi committed:
schemachanger: add column should enforce default value constraints

Previously, when adding a column we would not honor column
value constraints. This was inadequate because we could end
up violating any constraints when default value was not set.
For example if a column is not allowed to be null and no default
value is specified, then adding columns to a non-empty table should
fail. To address this, this patch will check if the table is not
empty and enforce these constraints.

Release note: None

fqazi committed:
sql/schemachanger: ensure ordering of added columns

Previously, when the declarative schema changer was adding
new columns, we didn't have any mechanism to enforce ordering.
As a result when multiple columns were added in a transaction,
they could be introduced in an arbitrary order based on the plan.
To address this, this patch sorts column additions based on
their IDs.

Release notes: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76983)
<!-- Reviewable:end -->

fqazi committed:
sql/schemachanger: block unsupported types in declarative schema changer

Previously, we incorrectly allowed support for generated
identities and columns with vector types during add column in the
declarative schema changer. This was inadequate because
our support for generated identities is incomplete, and
add columns with vector types is not allowed at all. To address
this, this patch will revert to the legacy schema changer for
generated identities and block adding columns of vector types.

Release note: None

fqazi committed:
sql/schemachanger: block certain default expression for add column

Previously, default expresions with sequence expressions were
allowed in the declarative schema changer when adding columns,
however if these expressions hit an error during backfill we
didn't surface appropriate errors. To address this, this patch
will disable the declarative schema changer when adding a column
requiring sequence expressions.

Release note: None

fqazi committed:
sql/schemachanger: fallback to legacy schema changer for computed column errors

Previously, when dealing with expressions for computed columns
there was a possibility that the computed expression was
referring to a newly added column. This could lead
to errors during the build phase, since our resolution
logic is not sophisticated inside the declarative schema
changer. To address this, we will temporarily fallback
to the legacy schema changer, whenever a parsing error
is hit with computed expressions.

Release note: None

fqazi committed:
sql/schemachanger: add missing telemetry for add column

Previously, the declarative schema changer was missing
telemetry for different types of computed expressions,
enum types, and type usage stats. This was inadequate,
since it would regress functionality. To address this,
this patch will introduce telemetry for computed expressions,
enum types, and type usage stats.

Release note: None

fqazi committed:
sql/schemachanger: support empty column family names

Previously, the declarative schema changer did not
properly support empty column family names. This led
to us updating the descriptor have entries with an
empty family name leading to validation error. To
address this, this patch will generate a family name
when an empty family name is specified.

Release note: None

fqazi committed:
sql/schemachanger: add column should detect system columns

Previously, the declarative schema changer did not
generate appropriate errors when attempting to
add column with the same name as a system column.
To address this, this patch will generate the appropriate
error when a column is added same name as a system column
within a table.

Release note: None

fqazi committed:
sql/schemachanger: enforce schema usage privileges

Previously, the declarative schema changer did not
properly enforce usage privileges on schemas, which
could allow add column operations in scenarios that
were previously disallowed. This was inadequate because
permissions were not properly enforced in this case.
To address this, this patch adds enforcement for
usage privileges on schemas.

Release note: None

fqazi committed:
sql/schemachanger: add event log entries for declarative schema changes

Previously, the declarative schema changer did not generate
event log entries for completion / reverts of complex
schema changes like column additions. This patch adds
support for generating these event log entries.

Release note: None

fqazi committed:
sql/schemachanger: statements are not correctly combined in CTEs

Previously, in certain cases like the geometry functions we
would internally execute multiple DDL statements under a CTE.
This led to persistence issues, where each statement would be
planned independently, leading to multiple column additions
failing when using the geometry builtins. To address this,
this patch add supports for replanning at execution time,
so that any missing statements are included in the currently
planned state.

Release note: None

fqazi committed:
sql/schemachanger: block implicit record types in expressions

Previously, the declarative schema changer did not
properly detect implicit record types properly, so
certain add column operations were not blocked properly.
This was inadequate because it could lead to scenarios,
where we were incorrectly accessing table descriptors as
type descriptors. To address this, this patch
will update the evaluate logic to parse expressions
appropriately and detect types which resolve to
tables.

Release note: None

fqazi committed:
sql/schemachanger: add column does not block cross db type references

Previously, the declarative schema changer did not detect
or check for cross database type references. This was
inadequate because functionality is blocked inside the
legacy schema changer already. To address this, this
patch will add logic to detect cross database reference
inside expressions and block them.

Release note: None

fqazi committed:
sql/schemachanger: disable declarative schema changer for certain tests

Previously, add column was not enabled by default with the declarative
schema changer, which does not support the new MVCC back filler fully
yet. This was inadequate, since we have a number of tests depending
on the new backfiller now, which only works with the legacy schema
changer. To address this, this patch will temporarily force the
legacy schema changer for these tests. Additionally, update
existing schame changer test related expected files.

Release note: None

fqazi committed:
sql/schemachanger: add stats refresh support for declarative schema changer

Previously, the declarative schema changer did not
properly invoke statistics refreshes when adding
columns. This was inadequate because statistics were
not properly recomputed for newly added columns. To
address this, this patch will add support for refreshing
stats as a new operation, which will be executed at the
end of the transaction. This operation is only invoked
for adding columns, but will be later expanded to other
changes.

Release note: None

fqazi committed:
sql/schemachanger: drop time was not populated when cleaning up indexes

Previously, the drop time was not populated when creating
index GC jobs in the declarative schema changer. This was
inadequate and would cause them to be GCed instantly leading
to unexpected behaviour. To address this, this patch starts
populating the drop times inside the declarative schema changer
for indexes.

Release note: None

fqazi committed:
sql/schemachanger: retry on declarative schema change errors

Previously, the declarative schema changer did not
support retries on schema change errors. This was
inadequate because if a retriable error was hit
during a schema change we would fail the schema
change and rollback. To address, this patch will
allow retrying declarative schema change when retriable
errors are hit.

Release note: None

fqazi committed:
sql/schemachanger: disable add column for regional by row partitioning

Previously, the declarative schema changer allowed adding
columns on tables multiregion tables with regional by row.
This was inadequate because when adding new columns in
the declarative schema changer we are not correctly computing
zone configs on indexes, which can lead to validation errors.
To address this, this patch will fall back to the legacy
schema change in these scenarios.

Release note: None

HonoreDB commented on pkg/ccl/changefeedccl/helpers_test.go:
```diff
@@ -315,7 +315,9 @@ SET CLUSTER SETTING kv.rangefeed.enabled = true;
 SET CLUSTER SETTING kv.closed_timestamp.target_duration = '1s';
 SET CLUSTER SETTING changefeed.experimental_poll_interval = '10ms';
 SET CLUSTER SETTING sql.defaults.vectorize=on;
+SET CLUSTER SETTING sql.defaults.use_declarative_schema_changer='off';
```
Would you mind adding a comment explaining which tests break with this on?

fqazi committed:
sql/schemachanger: support surfacing errors from backfiller

Previously, the declarative schema changer would wrap
all errors during execution to include plan related
information. This was inadequate because its valid
for user errors to be surfaced during execution, when
for example a backfill hits a division by zero error.
To address this, this patch introduces a new type of
error that will never be wrapped with plan information.

Release note: None

HonoreDB reviewed (commented):
Excited to see this! Two questions:

* I see you're turning off the declarative schema changer in changefeed tests. What breaks when add column uses the declarative schema changer? Will it be possible to remove that from the tests before merging this? (cdc team could make a PR to this branch if you need).
* You mention `Support empty column family names` as one of the fixes. cdc interacts with column family names--will we need to handle the possibility of a public column having an empty column family name? If so, is there a unique identifier we can use that can't conflict with other family names in the same table?

fqazi commented:
> Excited to see this! Two questions:
> 
> * I see you're turning off the declarative schema changer in changefeed tests. What breaks when add column uses the declarative schema changer? Will it be possible to remove that from the tests before merging this? (cdc team could make a PR to this branch if you need).

The current gap that we have is missing support for the newer backfiller inside the declarative schema changer, which is why the fallback was needed. The plan is to get that enabled sooner rather than later and remove any explicit disables we added for the declarative stuff. Let me go back and dig into these failures and document them

> * You mention `Support empty column family names` as one of the fixes. cdc interacts with column family names--will we need to handle the possibility of a public column having an empty column family name? If so, is there a unique identifier we can use that can't conflict with other family names in the same table?

Those get assigned an implicit name, the issue was that the declarative schema changer wasn't doing the right thing causing it to fail in general. They will have an implicit supplied name like today in the legacy schema changer.

fqazi committed:
changefeedccl: correctly distinguish pk swaps from truncates

Previously, the logic to detect table truncates vs primary
key swaps did not handle cases with the declarative schema
changer properly. This was inadequate because if a mutation
does not exist in the declarative world when adding column,
so any column additions appear as truncates. To address this,
this patch will check for any element manipulation on primary
indexes inside the declarative schema changer to distinguish
them from truncates.

Release note: None

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76983)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB and @samiskin)

___
*[`pkg/ccl/changefeedccl/helpers_test.go` line 318 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/76983#-N0YlRvelr80hO9pzBRR-r18-318:-N0frkzfDrdH6lF46Doy:bhj36rx) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c06f3f28e16ad493ca00a345c5cab4114cc0c98/pkg/ccl/changefeedccl/helpers_test.go#L318)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

Would you mind adding a comment explaining which tests break with this on?
</blockquote></details>

I modified this to do more targeted disables and found one issue with the schema change side of things. The remaining issues need test updates and are just linked to behavioural stuff between the two now, so temporarily left those tests on the legacy schema changer. We are going to address them with the issue that I have open.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
Seems to need a rebase and bazel generation. I'm interested in myself doing a fast follow to add the new index backfill protocol support. The changefeed behavior fixing will take a little more effort, I can take that on too. 

Do you want to merge this as-is or is it making you uncomfortable to walk back the index backfill protocol? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76983#-:-N19-qvF-dL2_nv43jJ0:b-1a940m)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB and @samiskin)



<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
I'm comfortable, assuming we are okay walking back temporarily. I'll dig into other missing bits on the side

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76983#-:-N19BTT8BjeICQiNHdtF:b-ynf94q)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB and @samiskin)



<!-- Sent from Reviewable.io -->

ajwerner commented on pkg/sql/schemachanger/scbuild/dependencies.go:
```diff
@@ -118,6 +121,13 @@ type CatalogReader interface {
+// checking their contents for example to determine if they are empty.
+type TableReader interface {
+	// IsTableEmpty returns if the table is empty.
+	IsTableEmpty(ctx context.Context, id descpb.ID, primaryIndexID descpb.IndexID) bool
```
don't you need to add this to the `sctestdeps` implementation?

ajwerner commented on pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go:
```diff
@@ -226,7 +226,7 @@ func init() {
 		),
 		element(scpb.Status_PUBLIC,
-			(*scpb.Column)(nil),
+			(*scpb.ColumnType)(nil),
```
what is this change about? It seems surprising

ajwerner commented on pkg/sql/schemachanger/scbuild/builder_state.go:
```diff
@@ -397,7 +397,9 @@ func (b *builderState) ComputedColumnExpression(
-		panic(err)
+		// This may be referencing newly added columns, so cheat and return
+		// a not implemented error.
+		panic(scerrors.NotImplementedErrorf(d, "computed column error %v", err))
```
this should use a wrap.

ajwerner commented on pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go:
```diff
@@ -469,7 +469,7 @@ func init() {
 		laterColNode, earlierColNode,
 		screl.MustQuery(
-			status.In(scpb.Status_DELETE_AND_WRITE_ONLY, scpb.Status_PUBLIC),
+			status.In(scpb.Status_WRITE_ONLY, scpb.Status_PUBLIC),
```
? we did this rename a while ago, yeah? bad rebase?

ajwerner commented on pkg/sql/schemachanger/scbuild/builder_state.go:
```diff
@@ -600,6 +609,8 @@ func (b *builderState) resolveRelation(
 	}
 	err, found := c.privileges[p.RequiredPrivilege]
 	if !found {
+		// Validate if this descriptor can be resolved under the current schema.
```
this smells to me. Shouldn't the `MayResolveTable` be taking this into consideration? Like if we don't have usage, then we need to look in the next schema. Consider:

As admin:
```sql
CREATE SCHEMA sc1;
CREATE SCHEMA sc2;
CREATE USER new_user;
GRANT CREATE, USAGE ON SCHEMA sc1 TO new_user;
GRANT CREATE, USAGE ON SCHEMA sc2 TO new_user;
```

```sql
CREATE TABLE sc1.foo (i INT PRIMARY KEY);
CREATE TABLE sc2.foo (i INT PRIMARY KEY);
```

admin
```sql
REVOKE CREATE, USAGE ON sc1 FROM new_user;
```

new_user:
```
SET search_path = sc1, sc2;
ALTER TABLE foo ADD COLUMN j INT;
```
The above should alter `sc2.foo`.

ajwerner replied:
cc @chengxiong-ruan as you've been playing here too

ajwerner commented on pkg/sql/opt/exec/execbuilder/testdata/sql_fn:
```diff
@@ -217,6 +205,19 @@ vectorized: true
               columns: ()
               estimated row count: 10 (missing stats)
 
+# It's ok to select other columns and multiple rows. Each geom column only gets
```
why did these need to move?

ajwerner commented on pkg/sql/schema_change_plan_node.go:
```diff
@@ -159,15 +165,50 @@ func (p *planner) waitForDescriptorSchemaChanges(
 	p := params.p
 	scs := p.ExtendedEvalContext().SchemaChangerState
+
+	// Our current state does not match what we planned, for this means that
```
nit: `for this`

ajwerner commented on pkg/sql/schemachanger/scexec/dependencies.go:
```diff
@@ -66,13 +66,18 @@ type Catalog interface {
 		event eventpb.EventPayload,
 	) error
+
+	// LogEventForSchemaChange write a schema cahnge event entry into the event log.
```
typo `cahnge`

ajwerner commented on pkg/sql/schemachanger/scbuild/builder_state.go:
```diff
@@ -319,6 +319,12 @@ func newTypeT(t *types.T) scpb.TypeT {
 // WrapExpression implements the scbuildstmt.TableHelpers interface.
 func (b *builderState) WrapExpression(expr tree.Expr) *scpb.Expression {
+	// Implicit table record references cannot be detected properly unless,
+	// we reparse the expression after serializing.
```
can you say more here?

ajwerner commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_column.go:
```diff
@@ -111,6 +112,22 @@ func alterTableAddColumn(
 
 	} else {
 		spec.colType.TypeT = b.ResolveTypeRef(d.Type)
+		// FIMXE:
```
?

ajwerner commented on pkg/sql/schemachanger/scexec/exec_mutation.go:
```diff
@@ -73,6 +73,10 @@ func executeDescriptorMutationOps(ctx context.Context, deps Dependencies, ops []
 	); err != nil {
 		return err
 	}
+	if err := refreshStatsForDescriptors(
+		ctx, mvs, deps.StatsRefresher()); err != nil {
```
nit: wrap this with a newline after the arguments too

ajwerner commented on pkg/sql/logictest/testdata/logic_test/alter_table:
```diff
@@ -774,7 +774,11 @@ CREATE TABLE default_err_test (foo text)
 statement error some_msg
+=======
+statement error pgcode foo some_msg
+>>>>>>> 3b0f9f04dc (sql/schemachanger: support surfacing errors from backfiller)
```
rebase detritus

ajwerner commented on pkg/sql/schemachanger/scerrors/errors.go:
```diff
@@ -90,3 +90,26 @@ func ConcurrentSchemaChangeDescID(err error) descpb.ID {
+
+// SchemaChangerUserError wraps an error as user consumable, which will surface
+// it from the declarative schema changer without any wrapping.
+func SchemaChangerUserError(err error) error {
```
I think this can use examples and some more words about what it would normally be wrapped with.

ajwerner commented on pkg/bench/rttanalysis/testdata/benchmark_expectations:
```diff
@@ -5,9 +5,9 @@ exp,benchmark
-13,AlterTableAddColumn/alter_table_add_1_column
-13,AlterTableAddColumn/alter_table_add_2_columns
-13,AlterTableAddColumn/alter_table_add_3_columns
+15,AlterTableAddColumn/alter_table_add_1_column
+16,AlterTableAddColumn/alter_table_add_2_columns
+17,AlterTableAddColumn/alter_table_add_3_columns
```
this is sad, do you know why we pick up the extra round-trips?

ajwerner reviewed (commented):
Close. So sorry it took me so darn long to look at this.

chengxiong-ruan commented:
*[`pkg/sql/schemachanger/scbuild/builder_state.go` line 612 at r69](https://reviewable.io/reviews/cockroachdb/cockroach/76983#-Mwg8flTTyJNLTVQJkzn-r69-612:-N19nbUS9ZTCnvVGKPwX:b3pss7x) ([raw file](https://github.com/cockroachdb/cockroach/blob/606abd0a935f468889e79cee27b19f853debcf27/pkg/sql/schemachanger/scbuild/builder_state.go#L612)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

cc @chengxiong-ruan as you've been playing here too 
</blockquote></details>

I feel like we should do something similar to `canResolveDescUnderSchema` planner for public. temp and virtual schemas... I ran into some backup tests failures where root user doesn't have `USAGE` privilege on some schema for some reason. I couldn't find the CI those test failed on though..damn...


<!-- Sent from Reviewable.io -->

chengxiong-ruan commented:
*[`pkg/sql/schemachanger/scbuild/builder_state.go` line 612 at r69](https://reviewable.io/reviews/cockroachdb/cockroach/76983#-Mwg8flTTyJNLTVQJkzn-r69-612:-N19rzPcE5wPcVPSLW50:b-b56vgt) ([raw file](https://github.com/cockroachdb/cockroach/blob/606abd0a935f468889e79cee27b19f853debcf27/pkg/sql/schemachanger/scbuild/builder_state.go#L612)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

I feel like we should do something similar to `canResolveDescUnderSchema` planner for public. temp and virtual schemas... I ran into some backup tests failures where root user doesn't have `USAGE` privilege on some schema for some reason. I couldn't find the CI those test failed on though..damn...
</blockquote></details>

I think we should fix the `search_path` schemas looping logic in general. Currently, as far as I can tell, if schema name is not given explicitly, we always use the first schema on `search_path` if we found something (table/index) we need, we never check schema privilege there :( this is definitely outside the scope of this PR.


<!-- Sent from Reviewable.io -->

chengxiong-ruan commented:
*[`pkg/sql/schemachanger/scbuild/builder_state.go` line 612 at r69](https://reviewable.io/reviews/cockroachdb/cockroach/76983#-Mwg8flTTyJNLTVQJkzn-r69-612:-N19tDgo5YBQRJo4hubE:b-l07pxe) ([raw file](https://github.com/cockroachdb/cockroach/blob/606abd0a935f468889e79cee27b19f853debcf27/pkg/sql/schemachanger/scbuild/builder_state.go#L612)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

I think we should fix the `search_path` schemas looping logic in general. Currently, as far as I can tell, if schema name is not given explicitly, we always use the first schema on `search_path` if we found something (table/index) we need, we never check schema privilege there :( this is definitely outside the scope of this PR.
</blockquote></details>

but that I meant something like this here: https://github.com/cockroachdb/cockroach/blob/c9e0194b19a03d55c6be92572aad3fbafc256334/pkg/sql/catalog/resolver/resolver.go#L444-L444

sorry for keep adding stuff here....


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76983)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @HonoreDB, and @samiskin)

___
*[`pkg/sql/schema_change_plan_node.go` line 185 at r70](https://reviewable.io/reviews/cockroachdb/cockroach/76983#-N0YlRvaezpVa9UHfKJk-r70-185:-N19nfjG2_X8VgJ8siTr:b3trvfn) ([raw file](https://github.com/cockroachdb/cockroach/blob/75c32e070c47cdb4b1f65d418dadac09ed2324a4/pkg/sql/schema_change_plan_node.go#L185)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: `for this`
</blockquote></details>

Changed this to:
	// Our current state does not match what was previously planned for, which means
	// that potentially we are running CTE's with ALTER statements. So, we are going
	// to re-plan the state to include the current statement since the statement
	// phase was not executed.
___
*[`pkg/sql/schemachanger/scbuild/builder_state.go` line 402 at r67](https://reviewable.io/reviews/cockroachdb/cockroach/76983#-Mwg8flTTyJNLTVQJkzn-r67-402:-N1A2Ajx7in2EoU74uLN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c3bd1b1ed88aa2afb494435f1609b3f3c3dbe09/pkg/sql/schemachanger/scbuild/builder_state.go#L402)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

this should use a wrap. 
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scbuild/builder_state.go` line 612 at r69](https://reviewable.io/reviews/cockroachdb/cockroach/76983#-Mwg8flTTyJNLTVQJkzn-r69-612:-N1ArB2oBXjsb6mQ0P_w:b-52pq2s) ([raw file](https://github.com/cockroachdb/cockroach/blob/606abd0a935f468889e79cee27b19f853debcf27/pkg/sql/schemachanger/scbuild/builder_state.go#L612)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

but that I meant something like this here: https://github.com/cockroachdb/cockroach/blob/c9e0194b19a03d55c6be92572aad3fbafc256334/pkg/sql/catalog/resolver/resolver.go#L444-L444

sorry for keep adding stuff here....
</blockquote></details>

I'm gonna add this to the ADD COLUMN limitation issue. I'll separate that list out and treat some of them as blockers for default enable.
___
*[`pkg/sql/schemachanger/scbuild/builder_state.go` line 323 at r72](https://reviewable.io/reviews/cockroachdb/cockroach/76983#-Mwg8flTTyJNLTVQJkzn-r72-323:-N1ANIss6ZRm-XmA1JKt:behwkno) ([raw file](https://github.com/cockroachdb/cockroach/blob/32f25e7d49d01d00d766720aaae00c8f51ab800e/pkg/sql/schemachanger/scbuild/builder_state.go#L323)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

can you say more here? 
</blockquote></details>

Added details for why something convoluted is needed here:	

// We will serialize and reparse the expression, so that type information
	// annotations are directly embedded inside, otherwise while parsing the
	// expression table record implicit types will not be correctly detected
	// by the TypeCollectorVisitor.
___
*[`pkg/sql/schemachanger/scbuild/dependencies.go` line 128 at r61](https://reviewable.io/reviews/cockroachdb/cockroach/76983#-Mwg8flTTyJNLTVQJkzo-r61-128:-N1ANnw06VSkrSvo7ixD:bywqerq) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3e6e2905bd430138e3d62739176f428d67a8c54/pkg/sql/schemachanger/scbuild/dependencies.go#L128)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

don't you need to add this to the `sctestdeps` implementation?
</blockquote></details>

Ugh, let me address this. It got split into a different commit :(

Done.



_Code quote:_
```Go
IsTableEmpty
```
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table.go` line 29 at r61](https://reviewable.io/reviews/cockroachdb/cockroach/76983#-Mwg8flTTyJNLTVQJkzp-r61-29:-N1AOlp9ENHiHtPhJvrI:b2y9r2r) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3e6e2905bd430138e3d62739176f428d67a8c54/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table.go#L29)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I guess, really, should we squash this before merging it? 
</blockquote></details>

Let me move it over to the last commit.

Done.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_column.go` line 115 at r73](https://reviewable.io/reviews/cockroachdb/cockroach/76983#-Mwg8flTTyJNLTVQJkzq-r73-115:-N1AVr1g3-hTDTYFA-4R:bxi20fy) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4e5284dc978aa0a2430e617940ab57c63c35d8e/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_column.go#L115)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

?
</blockquote></details>

This is gone, think something was wrong with my older push

Done. 
___
*[`pkg/sql/schemachanger/scerrors/errors.go` line 100 at r78](https://reviewable.io/reviews/cockroachdb/cockroach/76983#-N0aKXEhDP7DcCC-nVKg-r78-100:-N1ARjbbEjqI1pB4KVJW:b266d4h) ([raw file](https://github.com/cockroachdb/cockroach/blob/3941437f653709e34184c4b61a398c32ba5f5442/pkg/sql/schemachanger/scerrors/errors.go#L100)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I think this can use examples and some more words about what it would normally be wrapped with. 
</blockquote></details>

Updated this with:

// SchemaChangerUserError wraps an error as user consumable, which will surface
// it from the declarative schema changer without any wrapping. Normally errors
// from the declarative schema changer get wrapped with plan details inside
// DecorateErrorWithPlanDetails, but if the user for example specifies an
// expression that will raise an error during backfill, then we need to surface
// the error from the backfiller directly without wrapping (for example a
// default expression with a division by zero).
___
*[`pkg/sql/schemachanger/scexec/dependencies.go` line 77 at r71](https://reviewable.io/reviews/cockroachdb/cockroach/76983#-N0YlRvaezpVa9UHfKJn-r71-77:-N1APwu33mGARUmzbGRq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3f3c12d84919d34019cb383a9868b03a79ea3ea/pkg/sql/schemachanger/scexec/dependencies.go#L77)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

typo `cahnge`
</blockquote></details>

Done.



_Code quote:_
```Go
LogEventForSchemaChange
```
___
*[`pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go` line 229 at r65](https://reviewable.io/reviews/cockroachdb/cockroach/76983#-N0YlRvXJ3heAZp_5KSo-r65-229:-N1ASl0E2du0Sy73cgMV:bkbwlk9) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ebe2d02ff323446e55d9734abb57f2ad9224aa9/pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go#L229)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

what is this change about? It seems surprising
</blockquote></details>

We needed the column names set for generating family names inside SetAddedColumnType. So, needed to change the rule ordering a bit.
___
*[`pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go` line 472 at r68](https://reviewable.io/reviews/cockroachdb/cockroach/76983#-N0YlRvXJ3heAZp_5KSo-r68-472:-N1Ar7Ty57FLSTsBAP0U:b-2cysha) ([raw file](https://github.com/cockroachdb/cockroach/blob/4511993d14c81b5c7ee2313d1b177ab5f9366119/pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go#L472)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

? we did this rename a while ago, yeah? bad rebase? 
</blockquote></details>

Yes should be fixed now.
___
*[`pkg/bench/rttanalysis/testdata/benchmark_expectations` line 10 at r80](https://reviewable.io/reviews/cockroachdb/cockroach/76983#-N0YlRvelr80hO9pzBRP-r80-10:-N1Bw804BmW2HgNlmckj:b-4nl32e) ([raw file](https://github.com/cockroachdb/cockroach/blob/01982c56af153570b21372561432af448f0b525e/pkg/bench/rttanalysis/testdata/benchmark_expectations#L10)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

this is sad, do you know why we pick up the extra round-trips? 
</blockquote></details>

Let me investigate this. The initial bump is caused by extra looks up for the database + schema for some of our resolution logic. But the increasing trend is worrying with columns.
___
*[`pkg/sql/logictest/testdata/logic_test/alter_table` line 781 at r78](https://reviewable.io/reviews/cockroachdb/cockroach/76983#-N0YlRvelr80hO9pzBR_-r78-781:-N1AQo_MAo4PEX2QlC93:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3941437f653709e34184c4b61a398c32ba5f5442/pkg/sql/logictest/testdata/logic_test/alter_table#L781)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

rebase detritus
</blockquote></details>

Done.



_Code quote:_
```
statement error pgcode foo some_msg
```
___
*[`pkg/sql/opt/exec/execbuilder/testdata/sql_fn` line 208 at r70](https://reviewable.io/reviews/cockroachdb/cockroach/76983#-N0YlRvaezpVa9UHfKJj-r70-208:-N1AQwAD0aPFFKGL-CoM:buw70ax) ([raw file](https://github.com/cockroachdb/cockroach/blob/75c32e070c47cdb4b1f65d418dadac09ed2324a4/pkg/sql/opt/exec/execbuilder/testdata/sql_fn#L208)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

why did these need to move?
</blockquote></details>

We started doing the correct thing here, where the explain can properly detect duplicate columns now. So, if we really add the columns the explain will fail like it should now with the declarative schema changer.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76983)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @HonoreDB, and @samiskin)

___
*[`pkg/bench/rttanalysis/testdata/benchmark_expectations` line 10 at r80](https://reviewable.io/reviews/cockroachdb/cockroach/76983#-N0YlRvelr80hO9pzBRP-r80-10:-N1JgMLM6pkqkDR1iXhN:b-qrplit) ([raw file](https://github.com/cockroachdb/cockroach/blob/01982c56af153570b21372561432af448f0b525e/pkg/bench/rttanalysis/testdata/benchmark_expectations#L10)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Let me investigate this. The initial bump is caused by extra looks up for the database + schema for some of our resolution logic. But the increasing trend is worrying with columns.
</blockquote></details>

The per-column increase is due to event logging :facepalm:  We generate unnecessary entries in each column's current design. The legacy schema changer has one entry for the entire ALTER table, we need to combine them here.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> mod the rebase

> Reviewed 1 of 95 files at r21, 1 of 97 files at r61, 1 of 8 files at r106, 1 of 5 files at r108, 3 of 4 files at r113, 6 of 30 files at r117, 1 of 35 files at r120.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76983#-:-N2CDTF0-BJFHIxZ0SBr:bcwhr1c)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @HonoreDB, and @samiskin)



<!-- Sent from Reviewable.io -->

fqazi committed:
sql: hydrate types in cached uncommitted descriptor lookups

Previously, when looking up descriptors in the cached
uncommitted descriptors we would not hydrate type information,
which could lead to certain operations like add columns
inside the declarative schema changer failing. This change,
adds logic to hydrate table descriptors inside the uncommitted
descriptor cache when used for lookups. Additionally, synthetic
descriptors are skipped when hydrating types, since during execution
these can be injected to mark descriptors as dropped for the statement
phase.

Release note: None

chengxiong-ruan reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> the hydration change

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76983#-:-N2E1gDe496cFTT0bg9W:b93qhug)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @HonoreDB, and @samiskin)



<!-- Sent from Reviewable.io -->

postamar commented:
This last push seems to have triggered a lot more test failures but hopefully there's a simple explanation. In any case, nice work!

fqazi commented:
I had to modify the hydration logic to skip synthetic descriptors since we mark things as dropped during execution in the statement phase with synthetic descriptors. The alternative was to include dropped descriptors resolving types for hydration, but that feels wrong.

ajwerner commented:
Looks like there's still one descriptor resolution issue to squash

fqazi commented:
Ah, I need to handle offline descriptors here as well for hydration.

fqazi commented:
@ajwerner @chengxiong-ruan TFTR!

fqazi commented:
bors r+

postamar commented:
This is a huge step forward, great work!
